spec/requests/courses_spec.rb

```diff
@@ -22,95 +22,154 @@
   let(:subject) { FactoryBot.create(:subject) }
   let(:valid_attributes) { { name: "Biology 101" } }
   let(:invalid_attributes) { { name: nil } }
-
-  before { login_as(user, scope: :user) }
+  let(:login) { login_as(user, scope: :user) }
 
   describe "GET /show" do
-    it "renders a successful response" do
-      course = FactoryBot.create(:course)
-      get subject_course_url(subject, course)
-      expect(response).to be_successful
+    context "without logging in" do
+      it "redirects to users/sign_in page" do
+        course = FactoryBot.create(:course)
+        get subject_course_url(subject, course)
+        expect(response).to redirect_to("/users/sign_in")
+      end
+    end
+
+    context "with logging in" do
+      it "renders a successful response" do
+        login
+        course = FactoryBot.create(:course)
+        get subject_course_url(subject, course)
+        expect(response).to be_successful
+      end
     end
   end
 
   describe "GET /edit" do
-    it "render a successful response" do
-      course = FactoryBot.create(:course)
-      get edit_subject_course_url(subject, course)
-      expect(response).to be_successful
+    context "without logging in" do
+      it "redirects to users/sign_in page" do
+        course = FactoryBot.create(:course)
+        get edit_subject_course_url(subject, course)
+        expect(response).to redirect_to("/users/sign_in")
+      end
     end
-  end
 
-  describe "POST /create" do
-    context "with valid parameters" do
-      it "creates a new Course" do
-        expect do
-          post subject_courses_url(subject), params: { course: valid_attributes }
-        end.to change(Course, :count).by(1)
+    context "with logging in" do
+      it "render a successful response" do
+        login
+        course = FactoryBot.create(:course)
+        get edit_subject_course_url(subject, course)
+        expect(response).to be_successful
       end
+    end
+  end
 
-      it "redirects to the courses list" do
+  describe "POST /create" do
+    context "without logging in" do
+      it "redirects to users/sign_in page" do
         post subject_courses_url(subject), params: { course: valid_attributes }
-        expect(response).to redirect_to(subject_url(subject))
+        expect(response).to redirect_to("/users/sign_in")
       end
     end
 
-    context "with invalid parameters" do
-      it "does not create a new Course" do
-        expect do
-          post subject_courses_url(subject), params: { course: invalid_attributes }
-        end.to change(Course, :count).by(0)
+    context "with logging in" do
```

Comment:
suggestion: I would create two contexts out of this rather than increase the number of max nested groups. Deeply nested groups are difficult to read. Something like: `when logging in with valid parameters` and `when logging in with invalid parameters`